### ISSUE TITLE
k3s on Fedora 33 fail DNS from worker node pod

### ISSUE THREAD

**Issue by qlixed** [k3s-io/k3s#2935]
**Environmental Info:**
K3s Version: 
Master:
[root@pi3po1 ~]# k3s -v
k3s version v1.20.2+k3s1 (1d4adb03)
go version go1.15.5

Worker:
[root@pi3po2 ~]# k3s -v
k3s version v1.20.2+k3s1 (1d4adb03)
go version go1.15.5

Node(s) CPU architecture, OS, and Version: Aarch64, Fedora 33.

Master:
[root@pi3po1 ~]# uname -a
Linux pi3po1 5.10.15-200.fc33.aarch64 #1 SMP Wed Feb 10 17:33:15 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

Worker:
[root@pi3po2 ~]# uname -a
Linux pi3po2 5.10.15-200.fc33.aarch64 #1 SMP Wed Feb 10 17:33:15 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

Cluster Configuration:
1 master, 1 worker.

**Describe the bug:**
Cannot get DNS responses on a pod running in a worker node.
Master host config reference:
- Stub resolver disabled.
- Firewall config:
[root@pi3po1 ~]# firewall-cmd --state
inot running
[root@pi3po1 ~]# iptables --version
iptables v1.8.5 (legacy)
[root@pi3po1 ~]# iptables -L -n
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-EXTERNAL-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes externally-visible service portals */
KUBE-FIREWALL  all  --  0.0.0.0/0            0.0.0.0/0           

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
KUBE-FORWARD  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding rules */
KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes service portals */
KUBE-EXTERNAL-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes externally-visible service portals */
ACCEPT     all  --  10.42.0.0/16         0.0.0.0/0           
ACCEPT     all  --  0.0.0.0/0            10.42.0.0/16        

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes service portals */
KUBE-FIREWALL  all  --  0.0.0.0/0            0.0.0.0/0           

Chain KUBE-EXTERNAL-SERVICES (2 references)
target     prot opt source               destination         

Chain KUBE-FIREWALL (2 references)
target     prot opt source               destination         
DROP       all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
DROP       all  -- !127.0.0.0/8          127.0.0.0/8          /* block incoming localnet connections */ ! ctstate RELATED,ESTABLISHED,DNAT

Chain KUBE-FORWARD (1 references)
target     prot opt source               destination         
DROP       all  --  0.0.0.0/0            0.0.0.0/0            ctstate INVALID
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding rules */ mark match 0x4000/0x4000
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED

Chain KUBE-KUBELET-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-PROXY-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-SERVICES (2 references)
target     prot opt source               destination         

Worker host config reference:
- Stub resolver disabled.
- Firewall config:
[root@pi3po2 ~]# firewall-cmd --state
not running
[root@pi3po2 ~]# iptables --version
iptables v1.8.5 (legacy)
[root@pi3po2 ~]# iptables -L -n
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-FIREWALL  all  --  0.0.0.0/0            0.0.0.0/0           
KUBE-EXTERNAL-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes externally-visible service portals */

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
KUBE-FORWARD  all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding rules */
KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes service portals */
KUBE-EXTERNAL-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes externally-visible service portals */
ACCEPT     all  --  10.42.0.0/16         0.0.0.0/0           
ACCEPT     all  --  0.0.0.0/0            10.42.0.0/16        

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-FIREWALL  all  --  0.0.0.0/0            0.0.0.0/0           
KUBE-SERVICES  all  --  0.0.0.0/0            0.0.0.0/0            ctstate NEW /* kubernetes service portals */

Chain KUBE-EXTERNAL-SERVICES (2 references)
target     prot opt source               destination         

Chain KUBE-FIREWALL (2 references)
target     prot opt source               destination         
DROP       all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
DROP       all  -- !127.0.0.0/8          127.0.0.0/8          /* block incoming localnet connections */ ! ctstate RELATED,ESTABLISHED,DNAT

Chain KUBE-FORWARD (1 references)
target     prot opt source               destination         
DROP       all  --  0.0.0.0/0            0.0.0.0/0            ctstate INVALID
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding rules */ mark match 0x4000/0x4000
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
ACCEPT     all  --  0.0.0.0/0            0.0.0.0/0            /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED

Chain KUBE-KUBELET-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-PROXY-CANARY (0 references)
target     prot opt source               destination         

Chain KUBE-SERVICES (2 references)
target     prot opt source               destination

**Steps To Reproduce:**

1- Install fedora 33 minimal on AARCH64 (RPI3b+)
2- Install k3s server on a node
3- Test name resolution from a pod, it works ok when pod run in the master node.
4- Add an agent node (worker)
5- Test name resolution from a pod, it fails when the pod run on the worker node.

**Expected behavior:**

Name resolution should work all across the cluster.

**Actual behavior:**

Name resolution didn't work from the added worker node.

**Additional context / logs:**
If you require further information please let me know.

**Comment by clambin** [comment]
same behaviour on Ubuntu 21.04.

**Comment by brandond** [comment]
This is usually caused by something blocking communication between pods on one node trying to access a service whose pod is running on a different node. Ensure that you have disabled ufw/firewalld before starting K3s. You may need to disable the firewall and reboot the node in order for all the iptables rules to get recreated in the proper order.

**Comment by clambin** [comment]
no firewall here. from what I can tell, it works for pods that run on the same node as the coredns pod, but not for pods on other nodes.  Running k3s with "--flannel-backend=host-gw" solves that problem.

**Comment by brandond** [comment]
That indicates that your network infrastructure doesn't support vxlan, which is the default Flannel backend. What sort of network are you running your nodes on?

**Comment by clambin** [comment]
Not sure what you mean by "what sort of network"?   I'm running ubuntu 21.04, with a static IPv4 address.  Here's my Netplan:

```
network:
    version: 2
    ethernets:
        eth0:
            dhcp4: false
            addresses:
              - 192.168.0.11/24
            gateway4: 192.168.0.1
            nameservers:
              addresses: [ 8.8.8.8, 8.8.4.4 ]
```

Here's some vxlan-related entries from my syslog:

```
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.334676   37174 vxlan.go:121] VXLAN config: VNI=1 Port=0 GBP=false Learning=false DirectRouting=false
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.420511   37174 flannel.go:78] Wrote subnet file to /run/flannel/subnet.env
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.420599   37174 flannel.go:82] Running backend.
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.420671   37174 vxlan_network.go:60] watching for new subnet leases
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.433602   37174 iptables.go:145] Some iptables rules are missing; deleting and recreating rules
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.433686   37174 iptables.go:167] Deleting iptables rule: -s 10.42.0.0/16 -d 10.42.0.0/16 -j RETURN
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.433721   37174 iptables.go:145] Some iptables rules are missing; deleting and recreating rules
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.433762   37174 iptables.go:167] Deleting iptables rule: -s 10.42.0.0/16 -j ACCEPT
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.440494   37174 iptables.go:167] Deleting iptables rule: -s 10.42.0.0/16 ! -d 224.0.0.0/4 -j MASQUERADE --random-fully
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.440643   37174 iptables.go:167] Deleting iptables rule: -d 10.42.0.0/16 -j ACCEPT
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.446973   37174 iptables.go:167] Deleting iptables rule: ! -s 10.42.0.0/16 -d 10.42.3.0/24 -j RETURN
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.448145   37174 iptables.go:155] Adding iptables rule: -s 10.42.0.0/16 -j ACCEPT
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.452921   37174 iptables.go:167] Deleting iptables rule: ! -s 10.42.0.0/16 -d 10.42.0.0/16 -j MASQUERADE --random-fully
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.459786   37174 iptables.go:155] Adding iptables rule: -s 10.42.0.0/16 -d 10.42.0.0/16 -j RETURN
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.465114   37174 iptables.go:155] Adding iptables rule: -d 10.42.0.0/16 -j ACCEPT
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.472533   37174 iptables.go:155] Adding iptables rule: -s 10.42.0.0/16 ! -d 224.0.0.0/4 -j MASQUERADE --random-fully
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.484983   37174 iptables.go:155] Adding iptables rule: ! -s 10.42.0.0/16 -d 10.42.3.0/24 -j RETURN
Apr 28 19:59:51 worker1 k3s[37174]: I0428 19:59:51.497653   37174 iptables.go:155] Adding iptables rule: ! -s 10.42.0.0/16 -d 10.42.0.0/16 -j MASQUERADE --random-fully
```

Before I moved the master from Raspian 32bit to Ubuntu 64bit,  networking worked (mostly: sometimes DNS connectivity just stopped working and I had to restart k3s agent; there's a separate issue for that), so inherently,  the nodes do support vxlan.
